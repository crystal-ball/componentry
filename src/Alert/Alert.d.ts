import * as React from 'react'

interface AlertProps {
  active?: boolean,
  ariaTitle?: string,
  color?: string,
  deactivate?: Function,
  dismissible:? boolean,
  outline:? boolean,
}

/** https://crystal-ball.github.io/componentry/components/block */
declare const Alert: React.ComponentType<AlertProps>

export default Alert
