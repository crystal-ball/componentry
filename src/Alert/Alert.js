import React from 'react'
import element from '../element'
import { closeBase } from '../Close/Close'
import { useTheme } from '../Theme/Theme'
import staticComponent from '../factories/static-component'
import { useActive, useVisible } from '../hooks'

/**
 * @typedef {Object} Props
 * @property {string} Props.ariaTitle Set a specific aria title
 * @property {string} Props.color Set the theme color of the alert
 * @property {boolean} Props.dismissible If true the alert is dismissible
 * @property {boolean} Props.outline If true, the alert will have outline styles
 */

/**
 * [Alert component üìù](https://componentry.design/components/alert)
 */
export default function Alert(props) {
  /** @type {Props} */
  const {
    children,
    active: propsActive,
    ariaTitle,
    color,
    deactivate,
    dismissible,
    outline,
    ...rest
  } = { ...useTheme('Alert'), ...useActive(), ...props }

  const { active, visible } = useVisible(propsActive)

  return element({
    'role': 'alert',
    'componentCx': {
      'alert': true,
      'fade': dismissible, // Only include opacity transition class for disimissible alerts
      visible,
      [`alert-${color}`]: color,
      'alert-outline': outline,
    },
    // ‚ö†Ô∏è Only include aria-hidden value if the alert is dismissible
    'aria-hidden': dismissible ? String(!active) : undefined,
    'children': (
      <>
        {/* Provide the alert color context for screen readers */}
        <div className='sr-only'>{ariaTitle || `${color} alert`}</div>

        {/* Alert contents */}
        <div className='alert-content'>{children}</div>

        {/* Render a close button or null depending on configs */}
        {dismissible && (
          <Alert.Close
            className={`alert-close font-color-${color}`}
            onClick={deactivate}
          />
        )}
      </>
    ),
    ...rest,
  })
}
Alert.displayName = 'Alert'

/**
 * [Alert close component üìù](https://componentry.design/components/alert)
 */
Alert.Close = staticComponent('AlertClose', closeBase)
