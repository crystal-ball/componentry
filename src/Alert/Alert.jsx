// @flow
import React, { createElement } from 'react'
import { object, shape } from 'prop-types'
import classNames from 'classnames'

import Close from '../Close/Close'
import withActive from '../withActive/withActive'

import type { ThemeColors } from '../utils/theme'
import type { ActiveProps } from '../withActive/withActive'
import type { ElementProps } from '../component-factories/element-factory'

type Props = {
  /**
   * Provides contextual information to screen readers that sighted users would gain
   * from the alert color. Defaults to: 'COLOR alert', eg 'Success alert'
   */
  ariaTitle: string,
  /**
   * Theme color for alert
   */
  color: ThemeColors,
  /**
   * Pass `false` to suppress alert dismiss button
   */
  dismissible?: boolean,
  /**
   * Length of opacity transition, defaults to 300ms or `THEME` value if set using
   * `ThemeProvider`.
   */
  transitionDuration?: number,
} & ActiveProps &
  ElementProps

type Context = { [string]: { [string]: any } }

/**
 * Alerts provide contextual feedback to users. Alerts are available in the info
 * theme colors success, info, warning and danger. They are not available in primary
 * or secondary theme colors because they are intended to be used for alerting with
 * context. For non alert information blocks a card with theme color primary or
 * secondary can be used.
 */
const Alert = (props: Props, { THEME = {} }: Context) => {
  const componentCtx = THEME.Alert || {}
  const {
    ariaTitle,
    activate, // prevent dom inclusion
    active,
    as,
    children,
    className,
    color,
    dismissible,
    deactivate,
    visible,
    ...rest
  } = { ...componentCtx, ...props }

  return createElement(
    as || 'div',
    {
      role: 'alert',
      className: classNames('alert', 'fade', componentCtx.className, className, {
        [`alert-${color}`]: color,
        show: visible,
      }),
      // hidden state is updated after active opacity transition
      'aria-hidden': active ? 'false' : 'true',
      ...rest,
    },
    // Provide the alert color context for screen readers
    <div className="sr-only">{ariaTitle || `${color} alert`}</div>,
    // Alert contents:
    <div className="alert-content">{children}</div>,
    // Render a close button or null depending on configs
    dismissible && <Close onClick={deactivate} className={`text-${color}`} />,
  )
}

Alert.displayName = 'Alert'
Alert.defaultProps = {
  /* eslint-disable react/default-props-match-prop-types */
  active: true,
  visible: true,
  dismissible: false,
}
Alert.contextTypes = { THEME: shape({ Alert: object }) }

const withActiveAlert = withActive({ defaultActive: true, transitionState: true })(
  Alert,
)

export default withActiveAlert
