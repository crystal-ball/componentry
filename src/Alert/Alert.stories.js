import React from 'react'
import { storiesOf } from '@storybook/react'
import { boolean, select } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import Alert from './Alert'
import Active from '../Active/Active'
import { Anchor } from '../Type/Type'

const colors = {
  Primary: 'primary',
  Secondary: 'secondary',
  Info: 'info',
  Light: 'light',
  Dark: 'dark',
  Success: 'success',
  Warning: 'warning',
  Danger: 'danger',
}

storiesOf('Alert', module)
  .add('<Alert />', () => (
    <Active defaultActive>
      <Alert
        color={select('Color', colors, 'primary')}
        outline={boolean('Outline', false)}
        dismissible={boolean('Dismissible', true)}
      >
        <h4 className="alert-heading">Well done!</h4>
        <p className="mb-0">You successfully read this important alert message.</p>
        <hr className="hr" />
        <Anchor href="#" className="alert-link">
          Go home
        </Anchor>
      </Alert>
    </Active>
  ))
  .add('Collection', () => (
    <>
      <div>
        <h2 className="storybook-section-header">Colors</h2>
        <div className="d-flex flex-column">
          {Object.keys(colors).map(color => (
            <Active defaultActive key={color}>
              <Alert color={colors[color]} dismissible>
                {color} - Provide contextual info with an alert.
              </Alert>
            </Active>
          ))}
        </div>

        <h2 className="storybook-section-header">Outline</h2>
        <div className="d-flex flex-column">
          {Object.keys(colors).map(color => (
            <Active defaultActive key={color}>
              <Alert color={colors[color]} outline dismissible>
                {color} - Provide contextual info with an alert.
              </Alert>
            </Active>
          ))}
        </div>

        <h2 className="storybook-section-header">Aria title</h2>
        <Alert
          color="primary"
          ariaTitle="Alert provides context on setting the aria title"
        >
          Pass an <code>ariaTitle</code> to provide more context.
        </Alert>
      </div>
    </>
  ))
