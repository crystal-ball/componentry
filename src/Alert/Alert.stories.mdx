import {
  Meta,
  Title,
  Story,
  Preview,
  Props,
  Description,
} from '@storybook/addon-docs/blocks'
import { boolean, select } from '@storybook/addon-knobs'
import ElementsProvider from '../storybook/ElementsProvider'

import Active from '../Active/Active'
import Alert from './Alert'

<Meta title='Components|Alert' component={Alert} title='Alerts' />

<ElementsProvider>

# Alerts

<Description
  of={Alert}
  markdown='Provide contextual feedback messages for user actions using theme colors.'
></Description>

Alerts can contain any HTML elements, and can be dismissible or static. A dismissible
Alert requires Componentry Active `active` and `deactivate` props, which can be passed
directly or the Alert can be nested inside of an Active component to automatically create
an uncontrolled Alert.

<Alert color='primary' outline={boolean('outline', false)}>
  <h4 className='alert-heading'>Well done!</h4>
  <p className='mb-0'>You successfully read this important alert message.</p>
  <hr className='hr' />
  <a href='#' className='a alert-link'>
    Go home
  </a>
</Alert>

<Props of={Alert} />

You can also create outline alerts that have only a border-left

<Alert color='primary' outline>
  <h4 className='alert-heading'>Well done!</h4>
  <p className='mb-0'>You successfully read this important alert message.</p>
  <hr className='hr' />
  <a href='#' className='a alert-link'>
    Go home
  </a>
</Alert>

## Dismissible

Create a dismissible alert by wrapping it in an `Active` component.

<Active defaultActive>
  <Alert color='primary' dismissible>
    <h4 className='alert-heading'>Well done!</h4>
    <p className='mb-0'>You successfully read this important alert message.</p>
    <hr className='hr' />
    <Anchor href='#' className='alert-link'>
      Go home
    </Anchor>.
  </Alert>
</Active>

### Aria title

Provide an aria title for alerts with better accessibility.

<Alert color='primary' ariaTitle='Alert provides context on setting the aria title'>
  Pass an `ariaTitle` to provide more context.
</Alert>

</ElementsProvider>
