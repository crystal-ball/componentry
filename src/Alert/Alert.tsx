/* eslint-disable @typescript-eslint/no-empty-interface */

import React from 'react'
import { closeBase } from '../Close/Close'
import { useTheme } from '../Theme/Theme'
import { useActive, useVisible } from '../hooks'
import { ComponentBaseProps } from '../utils/types'
import { element } from '../utils/element-creator'
import { staticComponent } from '../utils/static-component-builder'

interface CardCloseProps extends ComponentBaseProps<'button'> {}

interface AlertProps extends ComponentBaseProps<'div'> {
  /** Sets a custom aria title */
  ariaTitle?: string
  /** Sets the theme color of the alert */
  color?: 'primary' | 'success' | 'warning' | 'critical'
  /** Deactivate handler called on click of Alert dismiss button */
  deactivate?: (event: React.MouseEvent<HTMLButtonElement>) => void
  /** Toggles alert dismissible feature */
  dismissible?: boolean
  /** Toggles alert outline styles */
  variant?: 'primary' | 'outline'
}

interface Alert {
  (props: AlertProps): React.ReactElement
  displayName: 'Alert'
  /**
   * [Alert close component üìù](https://componentry.design/components/alert)
   */
  Close: React.FC<CardCloseProps>
}

/**
 * [Alert component üìù](https://componentry.design/components/alert)
 */
export const Alert: Alert = (props) => {
  const {
    children,
    active: propsActive,
    ariaTitle,
    color,
    deactivate,
    dismissible,
    variant = 'primary',
    ...rest
  } = { ...useTheme<AlertProps>('Alert'), ...useActive(), ...props }

  const { active, visible } = useVisible(propsActive)

  return element('Alert', {
    'role': 'alert',
    'componentCx': [
      `alert-${variant}`,
      {
        [`alert-${color}`]: color,
        fade: dismissible, // Only include opacity transition class for disimissible alerts
        visible,
      },
    ],
    'aria-hidden': dismissible ? String(!active) : 'false',
    'children': (
      <>
        {/* Provide the alert color context for screen readers */}
        <div className='sr-only'>{ariaTitle || `${color} alert`}</div>

        {/* Alert contents */}
        <div className='alert-content'>{children}</div>

        {/* Render a close button or null depending on configs */}
        {dismissible && (
          <Alert.Close
            className={`alert-close font-color-${color}`}
            onClick={deactivate}
          />
        )}
      </>
    ),
    ...rest,
  })
}
Alert.displayName = 'Alert'

Alert.Close = staticComponent<CardCloseProps>('AlertClose', closeBase)
