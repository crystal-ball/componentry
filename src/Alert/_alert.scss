///                                      Componentry <Alert /> styles
/// -----------------------------------------------------------------

$alert: () !default;

$alert-defaults: (
  base: (
    // Make the alert container a flex container by default with space-between
    // to align close button to right side
    display: flex,
    align-items: flex-start,
    justify-content: space-between,
  ),
  // --- VARIANTS
  variants:
    (
      filled: (
        padding: theme-get('spacing.2'),
        border: 1px solid transparent,
        border-radius: theme-get('border-radius.md'),
      ),
    ),
);

// --------------------------------------------------------
// COLORS

$alert-colors: ();
$alert-theme-colors: ('primary', 'info', 'success', 'warning', 'error');
@each $color, $value in $alert-theme-colors {
  $alert-colors: map-assign(
    $alert-colors,
    (
      '.alert-filled.alert-color-#{$color}': (
        background: theme-get('colors.#{$color}.100'),
        border-color: theme-get('colors.#{$color}.300'),
        color: theme-get('colors.#{$color}.500'),
        '& .alert-link': (
          color: theme-get('colors.#{$color}.500'),
          font-weight: theme-get('font-weight.bold'),
        ),
        '& hr': (
          border-top-color: theme-get('colors.#{$color}.400'),
        ),
      ),
    )
  );
}

$alert-defaults: map-assign(
  $alert-defaults,
  (
    colors: $alert-colors,
  )
);

// --------------------------------------------------------
// CLASSES

$alert: map-assign($alert-defaults, $alert);

@include style-rule-from-map(componentry-class(alert), get($alert, 'base'));

@each $variant, $styles in get($alert, 'variants') {
  @include style-rule-from-map('.alert-#{$variant}', $styles);
}

@debug $alert-colors;
@include style-properties-from-map($alert-colors);

// --------------------------------------------------------
// ELEMENTS

.alert-content {
  flex-grow: 1;
  flex-shrink: 1;
}

.alert-heading {
  color: inherit;
  margin-bottom: theme-get('spacing.2');
}

.ðŸ…²-alert-close {
  flex-shrink: 0;
  margin-left: theme-get('spacing.1');

  // Animate close icon opacity on hover
  .icon-close {
    opacity: 0.6;
    transition: opacity 0.2s cubic-bezier(0.4, 0, 0.2, 1);

    &:hover,
    &:focus {
      opacity: 1;
    }
  }
}
