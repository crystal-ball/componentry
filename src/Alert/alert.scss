/* componentry/src/Alert/alert */

@mixin alert-variant($background, $border, $color) {
  color: $color;
  @include gradient-bg($background);
  border-color: $border;

  hr {
    border-top-color: darken($border, 5%);
  }

  .alert-link {
    color: darken($color, 10%);
  }
}

// Make the alert container a flex container by default with space-between to align
// close button to right side
.alert {
  align-items: flex-start;
  border: $alert-border-width solid transparent;
  display: flex;
  justify-content: space-between;
  margin-bottom: $alert-margin-bottom;
  padding: $alert-padding-y $alert-padding-x;
  position: relative;
  @include border-radius($alert-border-radius);
}

// Headings for larger alerts
.alert-heading {
  // Specified to prevent conflicts of changing $headings-color
  color: inherit;
}

// Provide class for links that match alerts
.alert-link {
  font-weight: $alert-link-font-weight;
}

// Alternate styles
//
// Generate contextual modifier classes for colorizing the alert.

@each $color, $value in $theme-colors {
  .alert-#{$color} {
    @include alert-variant(
      theme-color-level($color, $alert-bg-level),
      theme-color-level($color, $alert-border-level),
      theme-color-level($color, $alert-color-level)
    );
  }
}

// Outline styles
//
// Outline alerts have a transparent background with a bright left side border

@each $color, $value in $theme-colors {
  .alert-#{$color}.outline {
    background-color: transparent;
    border-color: $border-color;
    border-left: 5px solid $value;
  }
}
