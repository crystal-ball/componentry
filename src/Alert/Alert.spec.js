import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { Alert } from './Alert'

describe('<Alert/>', () => {
  // Basic library element test suite
  elementTests(Alert)
  elementTests(Alert.Close)

  test('should render an alert without a close button by default', () => {
    const { container } = render(<Alert color='success'>Warning!</Alert>)

    expect(container.firstChild).toHaveClass('ðŸ…²-alert alert-primary alert-color-success')
    expect(container.firstChild).toHaveAttribute('role', 'alert')
  })

  test('should not render a close button if not dismissible', () => {
    render(<Alert color='danger'>Warning!</Alert>)
    expect(screen.queryByLabelText('close')).toBeFalsy()
  })

  test('should bind passed deactivate to close button', async () => {
    const deactivate = jest.fn()
    const { container } = render(
      <Alert color='danger' deactivate={deactivate} active dismissible>
        Warning!
      </Alert>,
    )

    expect(screen.getByText('Warning!')).toBeInTheDocument()

    await fireEvent.click(screen.getByLabelText('close'))
    expect(deactivate).toHaveBeenCalled()

    // Alert visibility state change handler has been overridden, other than calling
    // passed deactivate Alert should still be visible & unchanved
    expect(container.firstChild).toHaveAttribute('aria-hidden', 'false')
  })

  // TODO: test active and deactivate handling
})

describe('<Alert /> snapshots', () => {
  test('renders correctly', () => {
    const { container } = render(
      <Alert color='danger' deactivate={jest.fn()} active dismissible>
        Warning!
      </Alert>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
