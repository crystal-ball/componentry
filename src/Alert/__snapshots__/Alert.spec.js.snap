// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> snapshots renders correctly 1`] = `
<div
  aria-hidden="false"
  class="ðŸ…²-alert alert-filled alert-color-danger fade"
  role="alert"
>
  <div
    class="sr-only"
  >
    danger alert
  </div>
  <div
    class="alert-content"
  >
    Warning!
  </div>
  <button
    aria-label="close"
    class="ðŸ…²-alert-close font-color-danger"
    type="button"
  >
    <svg
      class="ðŸ…²-icon icon-close font-variant"
      role="img"
    >
      <use
        href="#close"
        xlink:href="#close"
      />
    </svg>
  </button>
</div>
`;
